paddle/fluid/operators/detail/rpc_processor.h

```diff
@@ -0,0 +1,172 @@
+// limitations under the License.
+
+#pragma once
+
+#include <time.h>
+
+#include <functional>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "grpc++/grpc++.h"
+#include "grpc++/support/byte_buffer.h"
+#include "grpc++/support/slice.h"
+#include "grpc/support/log.h"
+#include "paddle/fluid/framework/blocking_queue.h"
+#include "paddle/fluid/framework/data_type.h"
+#include "paddle/fluid/framework/executor.h"
+#include "paddle/fluid/framework/lod_tensor.h"
+#include "paddle/fluid/framework/program_desc.h"
+#include "paddle/fluid/framework/scope.h"
+#include "paddle/fluid/framework/selected_rows.h"
+#include "paddle/fluid/framework/var_type.h"
+#include "paddle/fluid/operators/detail/sendrecvop_utils.h"
+#include "paddle/fluid/operators/detail/variable_response.h"
+
+namespace paddle {
+namespace operators {
+namespace detail {
+
+typedef std::pair<std::string, std::shared_ptr<VariableResponse>>
+    ReceivedMessage;
+
+class RPCProcessorCtx {
+ public:
+  RPCProcessorCtx()
+      : sync_mode_(true),
+        scope_(nullptr),
+        dev_ctx_(nullptr),
+        program_(nullptr),
+        executor_(nullptr) {}
+  virtual ~RPCProcessorCtx() {}
+
+  void SetSyncMode(bool sync_mode) { sync_mode_ = sync_mode; }
+  void SetScope(framework::Scope* scope) { scope_ = scope; }
+  void SetDevCtx(const platform::DeviceContext* dev_ctx) { dev_ctx_ = dev_ctx; }
+  void SetProgram(framework::ProgramDesc* program) { program_ = program; }
+  void SetExecutor(framework::Executor* executor) { executor_ = executor; }
+  void SetPrefetchPreparedCtx(
+      std::unique_ptr<framework::ExecutorPrepareContext> prepared) {
+    prefetch_ctx_.reset(prepared.release());
+  }
+
+  // Used for async
+  void SetGradToPreparedCtx(
+      std::unordered_map<
+          std::string, std::shared_ptr<framework::ExecutorPrepareContext>>* g) {
+    grad_to_prepared_ctx_ = g;
+  }
+
+  bool sync_mode() { return sync_mode_; }
+  framework::Scope* scope() { return scope_; }
+  const platform::DeviceContext* dev_ctx() { return dev_ctx_; }
+  framework::ExecutorPrepareContext* prefetch_ctx() {
+    return prefetch_ctx_.get();
+  }
+  framework::ProgramDesc* program() { return program_; }
+  framework::Executor* executor() { return executor_; }
+
+ protected:
+  bool sync_mode_;
+  framework::Scope* scope_;
+  const platform::DeviceContext* dev_ctx_;
+
+  std::unique_ptr<framework::ExecutorPrepareContext> prefetch_ctx_;
+  framework::ProgramDesc* program_;
+  framework::Executor* executor_;
+
+  // Used for async.
+  std::unordered_map<std::string,
+                     std::shared_ptr<framework::ExecutorPrepareContext>>*
+      grad_to_prepared_ctx_;
+};
+
+class GRPCProcessorCtx : public RPCProcessorCtx {
+ public:
+  GRPCProcessorCtx() : exit_flag_(false), fan_in_(-1) { clear_to_init(); }
+  virtual ~GRPCProcessorCtx() {}
+
+  bool ProcessSendImpl(const std::string& msg_name, framework::Variable* var,
```

Comment:
All handler functions can be saved in a `unordered_map<std::string, std::function>`, call the handler when RPC request arrives.